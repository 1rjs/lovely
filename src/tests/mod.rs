use std::fs;

use insta::glob;

use crate::{checker::Checker, ir::IRGenerator, lexer::Lexer, parser::Parser};

#[test]
fn compiler_tests() {
    glob!("source_files/*.lv", |path| {
        let input = fs::read_to_string(path).unwrap();

        // test the lexer
        let lexer = Lexer::new(&input);
        let tokens = lexer.collect::<Vec<_>>();
        insta::assert_debug_snapshot!(tokens);

        // test the parser
        let mut parser = Parser::new(&input);
        let ast = parser.parse().unwrap();
        insta::assert_debug_snapshot!(ast);

        // test the checker
        let mut checker = Checker::new();
        let checked_program = checker.check(&ast).unwrap();
        insta::assert_debug_snapshot!(checked_program);

        // test the IR
        let ir_generator = IRGenerator::new(checker.types);
        let ir = ir_generator.program_ir(&checked_program);
        let ir_string = ir
            .iter()
            .map(|b| b.to_string())
            .collect::<Vec<_>>()
            .join("\n");
        insta::assert_snapshot!(ir_string);
    });
}
