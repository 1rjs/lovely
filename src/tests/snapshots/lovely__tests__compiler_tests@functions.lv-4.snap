---
source: src/tests/mod.rs
expression: ir_string
input_file: src/tests/source_files/functions.lv
---
ENTRY:
  add#2 := FUN#0
  main#4 := FUN#2
  goto FUN#2

FUN#0:
  t1 := x#0 + y#1
  ret t1

FUN#2:
  t3 := add#2(4, 9)
  x#3 := t3
  exit x#3
